doc/refman/RefMan-ssr.tex

```diff
@@ -0,0 +1,4687 @@
+the type pattern \ssrN[2]{pattern}) are bound in \ssrN[2]{term}; \ssrN{ident} is
+also bound in \ssrN[3]{term} (but not in \ssrN[4]{term}), while the
+variables in \ssrN[1]{pattern} are bound only in \ssrN[3]{term}.
+
+\noindent
+Another variant allows to treat the else case first:
+
+\ssrC{if} \ssrN[1]{term} \ssrC{isn't} \ssrN[1]{pattern} \ssrC{then} \ssrN[2]{term} \ssrC{else} \ssrN[3]{term}
+
+Note that \ssrN[1]{pattern} eventually binds variables in \ssrN[3]{term}
+and not \ssrN[2]{term}.
+
+\subsection{Parametric polymorphism}\label{ssec:parampoly}
+Unlike ML, polymorphism in core Gallina is explicit: the type
+parameters of polymorphic  functions must be declared explicitly, and
+supplied at each point of use. However, \Coq{} provides two features
+to suppress redundant parameters:
+\begin{itemize}
+\item Sections are used to provide (possibly implicit) parameters for
+  a set of definitions.
+\item Implicit arguments declarations are used to tell \Coq{} to use
+  type inference to deduce some parameters from the context at each
+  point of call.
+\end{itemize}
+The combination of these features provides a fairly good emulation of ML-style
+polymorphism, but unfortunately this emulation breaks down for
+higher-order programming. Implicit arguments are indeed not inferred
+at all points of use, but only at
+points of call, leading to  expressions such as
+\begin{lstlisting}
+  Definition |*all_null*| (s : list T) := all (@null T) s.
+\end{lstlisting}
+Unfortunately, such higher-order expressions are quite frequent in
+representation functions, especially those which use \Coq{}'s
+\ssrC{Structure}s to emulate Haskell type classes.
+
+Therefore, \ssr{} provides a variant of \Coq{}'s implicit argument
+declaration, which  causes \Coq{} to fill in some implicit parameters
+at each point of use, e.g., the above definition can be written:
+\begin{lstlisting}
+  Definition |*all_null*| (s : list d) := all null s.
+\end{lstlisting}
+Better yet, it can be omitted entirely, since \ssrC{all_null s} isn't
+much of an improvement over \ssrC{all null s}.
+
+The syntax of the new declaration is
+
+\ssrC{Prenex Implicits} \ssrN{ident}$^+$.
+
+Let us denote $_1 \dots c_n$ the list of identifiers given to a
+\ssrC{Prenex Implicits} command.
+The command checks that each $c_i$ is the name of a functional
+constant, whose implicit arguments are prenex, i.e., the first $n_i >
+0$ arguments of $c_i$ are implicit; then it assigns
+\ssrC{Maximal Implicit} status to these arguments.
+
+As these prenex implicit arguments are ubiquitous and have often large
+display strings, it is strongly recommended to change the default
+display settings of \Coq{} so that they are not printed (except after a
+\ssrC{$\texttt{\textcolor{dkviolet}{Set }}$ Printing All} command).
+All \ssr{} library files thus start with the incantation
+\begin{lstlisting}
+  Set Implicit Arguments.
+  Unset Strict Implicit.
+  Unset Printing Implicit Defensive.
+\end{lstlisting}
```

Comment:
I wanted to ask what sense did this all make by opposition to:

    Set Implicit Arguments.
    Set Maximal Implicit Insertion.

Actually I tried to test it myself on the examples given in this section but the examples given here are completely non-reproducible. Please stop using `lstlisting` and use the environment for executing Coq code that the reference manual relies upon and check that the results do make sense. Do not open and close sections completely implicitly.